### ISSUE TITLE
Can't install from `dnf5-testing` COPR on fc40

### ISSUE THREAD

**Issue by lnicola** [rpm-software-management/dnf5#1580]
```
$ sudo dnf upgrade --refresh --allowerasing
==============================================================================================================================================================================================================================================
 Package                                     Architecture                           Version                                           Repository                                                                                         Size
==============================================================================================================================================================================================================================================
Installing group/module packages:
 dnf5                                        x86_64                                 5.2.4.0-1.fc40                                    copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf5-testing                                 653 k
     replacing  dnf.noarch 4.21.0-1.fc40
     replacing  yum.noarch 4.21.0-1.fc40
Installing dependencies:
 libdnf5                                     x86_64                                 5.2.4.0-1.fc40                                    copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf5-testing                                 1.0 M
 libdnf5-cli                                 x86_64                                 5.2.4.0-1.fc40                                    copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf5-testing                                 312 k

Transaction Summary
==============================================================================================================================================================================================================================================
Install  3 Packages

Total size: 2.0 M
Is this ok [y/N]: y
Downloading Packages:
[SKIPPED] dnf5-5.2.4.0-1.fc40.x86_64.rpm: Already downloaded                                                                                                                                                                                 
[SKIPPED] libdnf5-5.2.4.0-1.fc40.x86_64.rpm: Already downloaded                                                                                                                                                                              
[SKIPPED] libdnf5-cli-5.2.4.0-1.fc40.x86_64.rpm: Already downloaded                                                                                                                                                                          
Running transaction check
Transaction check succeeded.
Running transaction test
The downloaded packages were saved in cache until the next successful transaction.
You can remove cached packages by executing 'dnf clean packages'.
Error: Transaction test error:
  file /etc/dnf/dnf.conf from install of libdnf5-5.2.4.0-1.fc40.x86_64 conflicts with file from package dnf-data-4.21.0-1.fc40.noarch
  file /usr/share/man/man5/dnf.conf.5.gz from install of dnf5-5.2.4.0-1.fc40.x86_64 conflicts with file from package dnf-data-4.21.0-1.fc40.noarch
  file /usr/share/man/man8/dnf-download.8.gz from install of dnf5-5.2.4.0-1.fc40.x86_64 conflicts with file from package dnf-plugins-core-4.8.0-1.fc40.noarch
  file /usr/share/man/man8/dnf-system-upgrade.8.gz from install of dnf5-5.2.4.0-1.fc40.x86_64 conflicts with file from package dnf-plugins-core-4.8.0-1.fc40.noarch

$ sudo dnf swap dnf-data dnf5
Last metadata expiration check: 0:00:26 ago on Sb 13 iul 2024 15:10:43 +0300.
Error: 
 Problem: problem with installed package python3-dnf-4.21.0-1.fc40.noarch
  - package python3-dnf-4.19.0-1.fc40.noarch from fedora requires dnf-data = 4.19.0-1.fc40, but none of the providers can be installed
  - package python3-dnf-4.21.0-1.fc40.noarch from @System requires dnf-data = 4.21.0-1.fc40, but none of the providers can be installed
  - package python3-dnf-4.21.0-1.fc40.noarch from copr:copr.fedorainfracloud.org:rpmsoftwaremanagement:dnf5-testing requires dnf-data = 4.21.0-1.fc40, but none of the providers can be installed
  - package python3-dnf-4.21.0-1.fc40.noarch from updates requires dnf-data = 4.21.0-1.fc40, but none of the providers can be installed
  - conflicting requests
(try to add '--skip-broken' to skip uninstallable packages)
```

`dnf-unstable` still seems to work (it has `5.2.4.0-20240713005509.21.g9c2e2150.fc40` at this moment).

EDIT: it worked after `rpm -e --nodeps dnf-data dnf-plugins-core`

**Comment by jan-kolarik** [comment]
```
file /etc/dnf/dnf.conf from install of libdnf5-5.2.4.0-1.fc40.x86_64 conflicts with file from package dnf-data-4.21.0-1.fc40.noarch
```

This seems to be the main cause of the problem. And I believe one of the reasons is [this](https://github.com/rpm-software-management/dnf5/blame/5.2.5.0/dnf5.spec#L349) line we have in `dnf5.spec`. Looking at the [discussion](https://github.com/rpm-software-management/dnf5/pull/1443) where this change was introduced, it really tempts me to modify it to `Conflicts: dnf-data < 5` as suggested by @m-blaha back then. The problem is that without `dnf-data`, `python3-dnf` cannot be installed then as there is `Requires: %{name}-data = %{version}-%{release}`. Maybe we might change to require only `/etc/dnf/dnf.conf` by the `python3-dnf` package?

**Comment by evan-goode** [comment]
Could we add `Conflicts: /etc/dnf/dnf.conf` to `dnf5.spec`? That might force the solver to update `dnf-data` to the version provided by the Copr whenever `dnf5` is requested to be installed.

> Conflicting with dnf-data < 5 means dnf5 will build its own dnf-data subpackage. I guess we kept dnf-data subpackage in dnf4 to support uninstalling dnf5 without a need of downgrading dnf-data manually. (And of course because of the exact NVR dependency in python3-dnf). Is the downgrade still a concern?

A "seamless" downgrade of DNF 5 with dnf5_obsoletes_dnf down to DNF 4 is a low priority IMO; we can add uninstall instructions to the Copr if necessary. But moving dnf-data to dnf5 is a substantial change and it may not be worth it just to get this Copr working.

**Comment by evan-goode** [comment]
Here is a reproducer for the issue:

```
docker run -i --rm fedora:40 /bin/bash <<'EOF'
dnf update -y
dnf --version # Should be DNF 4.21
dnf install 'dnf-command(copr)' -y
dnf copr enable rpmsoftwaremanagement/dnf5-testing -y
dnf install dnf5 dnf5-plugins -y
EOF
```

If the system is already running dnf >= 4.21, installing `dnf5` will not update `dnf-data` to the version supplied by the Copr.

**Comment by evan-goode** [comment]
I managed to get this working without modifying any packages:

1. I set the "Mock bootstrap" setting on the Copr to "Disable". According to https://rpm-software-management.github.io/mock/Feature-bootstrap.html, Mock's bootstrap feature is useful when the build host's package manager is too old to directly install build dependencies into the buildroot. We don't need it in our case, and it's counterproductive because Mock tries to install DNF 5 from the Copr repo and fails because of the `/etc/dnf/dnf.conf` conflict. After disabling the Mock boostrap, packages in the Copr are now building again. 
2. I changed the installation instructions in the Copr description to say:

```
2. Jumping straight into installing DNF 5 won't work. First, you'll have to install the version of dnf-data provided by the Copr:

dnf reinstall --best dnf-data

3. Install dnf5:

dnf install dnf5
```

As the Copr description now says, `dnf install dnf5` still does not work, but IMO that's unimportant. We should not change too much in our upstream packages just to improve the UX of this "testing" Copr. As I understand it, this bug will not affect the F40->F41 upgrade since `dnf-data` will be upgraded from its `fc40` release to its `fc41` release in the same transaction `libdnf5` is installed.

Here is a script that verifies the Copr is working correctly:

```
docker run -i --rm registry.fedoraproject.org/fedora:40 /bin/bash -x <<'EOF'
dnf update -y
dnf --version # Should be DNF 4.21
dnf install 'dnf-command(copr)' -y
dnf copr enable rpmsoftwaremanagement/dnf5-testing -y
dnf reinstall --best dnf-data -y
dnf install dnf5 -y
dnf --version # Should be DNF 5
EOF
```
